torch/onnx/_internal/fx/registration.py

```diff
@@ -192,83 +79,122 @@ def _initiate_registry_from_torchlib(
+            for overload_func in aten_overloads_func.overloads:
+                symbolic_function = SymbolicFunction(
+                    onnx_function=overload_func, op_name=aten_name, is_custom=False
                 )
+                self._register(symbolic_function)
 
     @_beartype.beartype
-    def register(
-        self, name: str, opset: OpsetVersion, func: Callable, custom: bool = True
+    def _register(self, symbolic_function: SymbolicFunction) -> None:
+        """Registers a SymbolicFunction to an operator.
+
+        Args:
+            symbolic_function: The SymbolicFunction to register.
+        """
+        self._registry[symbolic_function.op_name].add(symbolic_function)
+
+    @_beartype.beartype
+    def register_custom_op(
+        self,
+        function: Union["onnxscript.OnnxFunction", "onnxscript.TracedOnnxFunction"],
+        domain: str,
+        op_name: str,
+        overload: Optional[str] = None,
     ) -> None:
-        """Registers overloaded functions to an ATen operator (overload).
+        """Registers a custom operator: torch.ops.<domain>.<op_name>.<overload>.
 
         Args:
-            name: The qualified name of the function to register. In the form of 'domain::op'.
-                E.g. 'aten::add' or 'aten::pow.int'.
-            opset: The opset version of the function to register.
-            func: The symbolic function to register.
-            custom: Whether the function is a custom function that overrides existing ones.
+            function: The onnx-sctip function to register.
+            domain: The domain of the operator to register.
+            op_name: The name of the operator to register.
+            overload: The overload of the operator to register. If it's default overload,
+                leave it to None.
 
         Raises:
-            ValueError: If the separator '::' is not in the name.
+            ValueError: If the name is not in the form of 'domain::op'.
         """
-        if "::" not in name:
-            raise ValueError(
-                f"The name must be in the form of 'domain::op', not '{name}'"
-            )
-        symbolic_functions = self._registry.setdefault(
-            name, _SymbolicFunctionGroup(name)
-        )
-        if custom:
-            symbolic_functions.add_custom(func, opset)
+        if overload is None:
+            # NOTE: "default" overload
+            name = f"{domain}::{op_name}"
         else:
-            symbolic_functions.add(func, opset)
+            name = f"{domain}::{op_name}.{overload}"
+        symbolic_function = SymbolicFunction(
+            onnx_function=function, op_name=name, is_custom=True
+        )
+        self._register(symbolic_function)
 
     @_beartype.beartype
-    def unregister(self, name: str, opset: OpsetVersion) -> None:
-        """Unregisters overloaded functions to an ATen operator (overload).
+    def get_functions(
+        self, domain: str, op_name: str, overload: Optional[str] = None
```

Comment:
I will write a dataclass. Credit: @justinchuby

Reply:
Created OpName dataclass in registration.

Reply:
Why do we need to distinguish a custom op from a regular op? This is extra information is not needed.

In theory, we should 1) be able to replace a buggy built-in (existing) op or 2) add a new (custom) operator. In both cases, the API should be the same for registering it or for checking whether they are registered.

When `register_op(domain, name, overload)` is called, it can overwrite any existing or custom function there was in the registry already. The overriding of a function by another is not an issue because during an export, we should not export the model with different versions of the same (domain, name, overload) anyway. Having extra ifnormation about custom op creates bug as Bowen pointed out above and make the code unnecessarily more complex.

Reply:
What happens when a user wants to register two implementations to the same op for different dtypes? Will overwriting cause any issues?

Reply:
I think @BowenBao was talking about the op naming, not the register_custom_op itself. The naming should be fixed by dataclass `OpName`.

> When register_op(domain, name, overload) is called, it can overwrite any existing or custom function there was in the registry already. The overriding of a function by another is not an issue because during an export, we should not export the model with different versions of the same (domain, name, overload) anyway. Having extra ifnormation about custom op creates bug as Bowen pointed out above and make the code unnecessarily more complex.

To actually override a specific onnx overload, we would need to introduce `OpSchemaWrapper` here to identify which one the user is trying to override with. Adding `custom` attribute avoids that complexity by simply adding custom function as overload candidate, and will be prioritized in dispatching. onnx overload is not like ATen overload that we can just look up by its "ID". They are identified by their OpSchema.

One could argue that users wouldn't be able to find out if he/she writes a wrong OpSchema until exporting (SARIF would tell them what happened.), as he/she couldn't observe that by `get_functions` because there is no deduplicate OpSchema in registry. However, this is a rare case we are talking about, and also I feel it's reasonable to expect them to write the correct OpSchema, if their plans were to "override the default one".

> What happens when a user wants to register two implementations to the same op for different dtypes? Will this happen? Will overwriting cause any issues?

I can't think of any concrete case, but I am assuming it would happen. In current implementation, it all depends on OpSchema matching, if you are perfect match/nearest match, then you are up.